torchsde/_core/methods/reversible_heun.py

```diff
@@ -0,0 +1,117 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import torch
+
+from .. import adjoint_sde
+from .. import base_solver
+from .. import misc
+from ...settings import SDE_TYPES, NOISE_TYPES, LEVY_AREA_APPROXIMATIONS, METHODS, METHOD_OPTIONS
+
+
+class ReversibleHeun(base_solver.BaseSDESolver):
+    weak_order = 1.0
+    sde_type = SDE_TYPES.stratonovich
+    noise_types = NOISE_TYPES.all()
+    levy_area_approximations = LEVY_AREA_APPROXIMATIONS.all()
+
+    def __init__(self, sde, **kwargs):
+        self.strong_order = 1.0 if sde.noise_type == NOISE_TYPES.additive else 0.5
+        super(ReversibleHeun, self).__init__(sde=sde, **kwargs)
+
+    def init_extra_solver_state(self, t0, y0):
+        return self.sde.f_and_g(t0, y0) + (y0,)
+
+    def step(self, t0, t1, y0, extra0):
+        f0, g0, z0 = extra0
+        dt = t1 - t0
+        dW = self.bm(t0, t1)
+
+        z1 = 2 * y0 - z0 + f0 * dt + self.sde.prod(g0, dW)
```

Comment:
Could we document what z here is for others, even though it seems obvious now after reading the paper.